File: example05ui/src/main/java/foo/bar/example/asafui/api/fruits/FruitsCustomError.java
package foo.bar.example.asafui.api.fruits;

import com.google.gson.annotations.SerializedName;

import co.early.asaf.core.Affirm;
import co.early.asaf.retrofit.MessageProvider;
import foo.bar.example.asafui.message.UserMessage;

import static foo.bar.example.asafui.message.UserMessage.ERROR_MISC;

/**
 *
 * <Code>
 *
 *  The server returns custom error codes in this form:
 *
 *  {
 *    "errorCode":"FRUIT_USER_LOCKED"
 *  }
 *
 * </Code>
 *
 *
 */
public class FruitsCustomError implements MessageProvider<UserMessage> {

    private ErrorCode errorCode;

    public FruitsCustomError(ErrorCode errorCode) {
        this.errorCode = errorCode;
    }

    public enum ErrorCode {

        @SerializedName("FRUIT_USER_LOGIN_CREDENTIALS_INCORRECT")
        LOGIN_CREDENTIALS_INCORRECT(UserMessage.ERROR_FRUIT_USER_LOGIN_CREDENTIALS_INCORRECT),
        @SerializedName("FRUIT_USER_LOCKED")
        USER_LOCKED(UserMessage.ERROR_FRUIT_USER_LOCKED),
        @SerializedName("FRUIT_USER_NOT_ENABLED")
        USER_NOT_ENABLED(UserMessage.ERROR_FRUIT_USER_NOT_ENABLED);

        public final UserMessage userMessage;

        ErrorCode(UserMessage userMessage) {
            this.userMessage = Affirm.notNull(userMessage);
        }
    }

    public UserMessage getMessage() {
        if (errorCode == null) {
            return ERROR_MISC;
        } else {
            return errorCode.userMessage;
        }
    }

}

File: app-examples/example-kt-08ktor/src/main/java/foo/bar/example/forektorkt/api/fruits/FruitsCustomError.kt
package foo.bar.example.forektorkt.api.fruits

import co.early.fore.net.MessageProvider
import foo.bar.example.forektorkt.message.ErrorMessage
import foo.bar.example.forektorkt.message.ErrorMessage.ERROR_MISC
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 *
 * <Code>
 *
 * The server returns custom error codes in this form:
 *
 * {
 * "errorCode":"FRUIT_USER_LOCKED"
 * }
 *
 * </Code>
 *
 * example http://www.mocky.io/v2/59ef2a6c2e00002a1a1c5dea
 *
 */
@Serializable
class FruitsCustomError(private val errorCode: ErrorCode?) : MessageProvider<ErrorMessage> {

    @Serializable
    enum class ErrorCode constructor(val errorMessage: ErrorMessage) {

        @SerialName("FRUIT_USER_LOGIN_CREDENTIALS_INCORRECT")
        LOGIN_CREDENTIALS_INCORRECT(ErrorMessage.ERROR_FRUIT_USER_LOGIN_CREDENTIALS_INCORRECT),

        @SerialName("FRUIT_USER_LOCKED")
        USER_LOCKED(ErrorMessage.ERROR_FRUIT_USER_LOCKED),

        @SerialName("FRUIT_USER_NOT_ENABLED")
        USER_NOT_ENABLED(ErrorMessage.ERROR_FRUIT_USER_NOT_ENABLED);

    }

    override val message: ErrorMessage = errorCode?.errorMessage ?: ERROR_MISC
}
